// typeorm-extension/dist/database/utils/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildDatabaseDropContext = exports.buildDatabaseCreateContext = void 0;
const data_source_1 = require("../../data-source");
async function setDatabaseContextOptions(context) {
    if (!context.options) {
        const dataSource = await (0, data_source_1.findDataSource)(context.findOptions);
        if (dataSource) {
            context.options = dataSource.options;
        }
        if (!context.options) {
            context.options = await (0, data_source_1.buildDataSourceOptions)();
        }
    }
    Object.assign(context.options, {
        subscribers: [],
        synchronize: false,
        migrationsRun: false,
        dropSchema: false,
    });
    return context;
}
async function buildDatabaseCreateContext(context) {
    context = context || {};
    context = await setDatabaseContextOptions(context);
    if (typeof context.synchronize === 'undefined') {
        context.synchronize = true;
    }
    if (typeof context.ifNotExist === 'undefined') {
        context.ifNotExist = true;
    }
    return context;
}
exports.buildDatabaseCreateContext = buildDatabaseCreateContext;
async function buildDatabaseDropContext(context) {
    context = context || {};
    context = await setDatabaseContextOptions(context);
    if (typeof context.ifExist === 'undefined') {
        context.ifExist = true;
    }
    return context;
}
exports.buildDatabaseDropContext = buildDatabaseDropContext;
